packages/orm/src/relations/Reference.ts

```diff
@@ -1,66 +1,60 @@
 import { Entity, IdOf } from "../EntityManager";
 import { CustomReference } from "./CustomReference";
-import { ManyToOneReference } from "./ManyToOneReference";
-import { OneToOneReference } from "./OneToOneReference";
+import { ManyToOneReferenceImpl } from "./ManyToOneReference";
+import { OneToOneReferenceImpl } from "./OneToOneReference";
 import { PolymorphicReference } from "./PolymorphicReference";
-import { Relation } from "./Relation";
+import { Relation, RelationT, RelationU } from "./Relation";
 
-const H = Symbol();
+// Exported and used in sub-interfaces/types per https://stackoverflow.com/a/70437874/355031
+export const ReferenceN = Symbol();
+
+// We could go back to `N extends never | undefined` once https://github.com/microsoft/TypeScript/issues/47213
+// is resolved so that `MarkLoaded` can keep matching on `Reference` and it works with sub-interfaces.
+export type MaybeUndefined<T, N extends NullOrNotNull> = N extends "null" ? T | undefined : T;
+export type NullOrNotNull = "null" | "not-null";
```

Comment:
```suggestion
export type Nullable = "nullable"
export type NotNullable = "not-nullable"
export type MaybeNullable = Nullable | NotNullable
```

then you could do just `MaybeNullable` instead of `NullOrNotNull`.  Less concise, but seems like it'd show intent better and not require hardcoded strings all over.